lib/Sema/TypeCheckDecl.cpp

```diff
@@ -2504,7 +2504,25 @@ NamingPatternRequest::evaluate(Evaluator &evaluator, VarDecl *VD) const {
 
   if (!namingPattern) {
     // Try type checking parent control statement.
-    if (auto parentStmt = VD->getParentPatternStmt()) {
+    if (auto condStmt = dyn_cast_or_null<LabeledConditionalStmt>(
+            VD->getParentPatternStmt())) {
+      // The VarDecl is defined inside a condition of a `if` or `while` stmt.
+      // Only type check the condition we care about
+      for (auto &cond : condStmt->getCond()) {
+        if (auto pat = cond.getPattern()) {
+          if (pat->containsVarDecl(VD)) {
+            // We don't care about isFalsable
+            bool isFalsable = false;
+            TypeChecker::typeCheckStmtConditionElement(cond, isFalsable,
+                                                       VD->getDeclContext());
+            // We have found the condition. No need to look at the other
+            // conditions
+            break;
+          }
+        }
```

Comment:
Don't we need this?
```
namingPattern = VD->getCanonicalVarDecl()->NamingPattern;
```
I was thinking something like:
```cpp
if (auto parentStmt = VD->getParentPatternStmt()) {
  if (auto condStmt = dyn_cast_or_null<LabeledConditionalStmt>(parentStmt)) {
     // Only type check the condition we care about
     for (auto &cond : condStmt->getCond()) {
        ....
     }
  } else {
    // We have some other parent stmt. Type check it completely.
    if (auto CS = dyn_cast<CaseStmt>(parentStmt))
      parentStmt = CS->getParentStmt();
    ASTNode node(parentStmt);
    TypeChecker::typeCheckASTNode(node, VD->getDeclContext()
  }
   
  namingPattern = VD->getCanonicalVarDecl()->NamingPattern;
}
```

Reply:
I thought I could do two things simultaneously. Reality proves me wrong again 😬.